compiler/noirc_frontend/src/hir/resolution/path_resolver.rs

```diff
@@ -55,17 +56,41 @@ pub fn resolve_path(
     def_maps: &BTreeMap<CrateId, CrateDefMap>,
     module_id: ModuleId,
     path: Path,
-) -> Result<ModuleDefId, PathResolutionError> {
+) -> Result<(ModuleDefId, Option<PathResolutionError>), PathResolutionError> {
     // lets package up the path into an ImportDirective and resolve it using that
+    let last_path_segment = path.last_segment();
     let import =
         ImportDirective { module_id: module_id.local_id, path, alias: None, is_prelude: false };
     let allow_referencing_contracts =
         allow_referencing_contracts(def_maps, module_id.krate, module_id.local_id);
 
-    let def_map = &def_maps[&module_id.krate];
-    let ns = resolve_path_to_ns(&import, def_map, def_maps, allow_referencing_contracts)?;
+    let (resolved_module, ns, mut warning) = resolve_path_to_ns(
+        &import,
+        module_id.krate,
+        module_id.krate,
+        def_maps,
+        allow_referencing_contracts,
+    )?;
 
-    let function = ns.values.map(|(id, _, _)| id);
-    let id = function.or_else(|| ns.types.map(|(id, _, _)| id));
-    Ok(id.expect("Found empty namespace"))
+    let (id, visibility) = ns
+        .values
+        .or(ns.types)
+        .map(|(id, visibility, _)| (id, visibility))
+        .expect("Found empty namespace");
+
+    warning = warning.or_else(|| {
+        if can_reference_module_id(
+            def_maps,
+            module_id.krate,
+            module_id.local_id,
+            resolved_module,
+            visibility,
+        ) {
+            None
+        } else {
+            Some(PathResolutionError::Private(last_path_segment.clone()))
+        }
+    });
```

Comment:
Looks like mostly duplicated code from `import.rs`, can you refactor it to a common helper function?